File: src/lib/finance/byTax.ts
/**
 * Belarus-specific tax calculators. All percentages and figures reflect the
 * mainstream rules as of 2024-2025; values that change every year (МЗП,
 * базовая величина, льготные пороги вычетов) are exposed as constants and
 * can be tweaked from one place.
 *
 * NOTE: this is an estimator — not a substitute for an accountant.
 */

/**
 * Год-зависимые параметры НК РБ. Значения МЗП, базовой величины и пороги
 * вычетов меняются почти каждый год — держим их в таблице по годам, чтобы
 * обновление сводилось к добавлению одной записи.
 *
 * ⚠️ Подтверждены только значения 2024 года. Значения 2025/2026 нужно
 * добавить из официальных источников (постановления Совмина по МЗП/БВ и
 * ст. 209 НК РБ по вычетам).
 */
export interface ByTaxYearConstants {
  /** Базовая величина (BV). */
  baseValueByn: number;
  /** Минимальная заработная плата (МЗП). */
  minWageByn: number;
  /** Стандартный вычет «на себя» (ст. 209 НК РБ) — порог дохода. */
  stdDeductionIncomeLimit: number;
  /** Сумма стандартного вычета «на себя». */
  stdDeduction: number;
  /** Вычет на одного ребёнка / иждивенца. */
  childDeduction: number;
  /** Вычет на ребёнка для семей с 2+ детьми / одинокого родителя. */
  childDeductionTwoPlus: number;
  /** Подоходный налог, %. */
  incomeTaxPct: number;
  /** ФСЗН с работника, %. */
  fsznEmployeePct: number;
}

export const BY_TAX_BY_YEAR: Record<number, ByTaxYearConstants> = {
  2024: {
    baseValueByn: 42,
    minWageByn: 626,
    stdDeductionIncomeLimit: 1054,
    stdDeduction: 174,
    childDeduction: 51,
    childDeductionTwoPlus: 97,
    incomeTaxPct: 13,
    fsznEmployeePct: 1,
  },
};

/** Последний год с подтверждёнными значениями — используется по умолчанию. */
export const CURRENT_TAX_YEAR = 2024;

/**
 * Возвращает константы для года: точное совпадение → ближайший предыдущий
 * заполненный год → последний доступный. Так калькулятор не падает, если
 * запрошен год, для которого таблица ещё не обновлена.
 */
export function getByTaxConstants(year: number = CURRENT_TAX_YEAR): ByTaxYearConstants {
  const years = Object.keys(BY_TAX_BY_YEAR)
    .map(Number)
    .sort((a, b) => a - b);
  if (BY_TAX_BY_YEAR[year]) return BY_TAX_BY_YEAR[year];
  const fallbackYear =
    [...years].reverse().find(y => y <= year) ?? years[years.length - 1];
  return BY_TAX_BY_YEAR[fallbackYear];
}

const CURRENT = getByTaxConstants(CURRENT_TAX_YEAR);

/** Базовая величина (BV) для текущего года. */
export const BASE_VALUE_BYN = CURRENT.baseValueByn;

/** Минимальная заработная плата (МЗП) для текущего года. */
export const MIN_WAGE_BYN = CURRENT.minWageByn;

/** Стандартный вычет «на себя» (ст. 209 НК РБ) — порог дохода. */
export const STD_DEDUCTION_INCOME_LIMIT = CURRENT.stdDeductionIncomeLimit;
export const STD_DEDUCTION = CURRENT.stdDeduction;

/** Вычет на ребёнка / иждивенца (по умолчанию). */
export const CHILD_DEDUCTION = CURRENT.childDeduction;
export const CHILD_DEDUCTION_TWO_PLUS = CURRENT.childDeductionTwoPlus;

/** Подоходный налог. */
export const INCOME_TAX_PCT = CURRENT.incomeTaxPct;

/** ФСЗН с работника (1%). */
export const FSZN_EMPLOYEE_PCT = CURRENT.fsznEmployeePct;

/**
 * Дополнительное удержание из зарплаты — профсоюз, ДМС, благотворительность,
 * пенсионная программа, кредитные удержания, алименты и т.д.
 */
export interface SalaryDeduction {
  /** Стабильный идентификатор. */
  id: string;
  /** Название (например «Профсоюз»). */
  label: string;
  /** Тип значения: процент от грязной или фикс. сумма в BYN. */
  kind: 'percent' | 'fixed';
  /** Значение: либо процент (0..100), либо BYN. */
  value: number;
  /**
   * Если `true` — удержание уменьшает налогооблагаемую базу (как соц. вычет).
   * По умолчанию `false` — удерживается из «на руки» после налогов.
   *
   * Профсоюзные взносы по сложившейся практике не уменьшают подоходный
   * налог, поэтому правильное значение для профсоюза — `false`.
   */
  taxable?: boolean;
}

export interface SalaryInput {
  /** Грязная зарплата (до налогов). */
  gross: number;
  /** Сколько детей у работника (для детского вычета). */
  children?: number;
  /** Иждивенцы (родители, неработающий супруг и т.д.). */
  dependents?: number;
  /** Применять стандартный вычет «на себя», если доход ≤ порога? */
  applyStandardDeduction?: boolean;
  /** Произвольные удержания: профсоюз, ДМС, благотворительность и т.д. */
  extraDeductions?: SalaryDeduction[];
}

export interface SalaryDeductionLine {
  id: string;
  label: string;
  amount: number;
  /** Уменьшила ли налоговую базу. */
  taxable: boolean;
}

export interface SalaryResult {
  gross: number;
  fszn: number;
  taxableBase: number;
  /** Сумма стандартных + детских + иждивенских вычетов. */
  deductions: number;
  /** Сумма pre-tax удержаний (уменьшили базу), напр. ДМС. */
  pretaxDeductions: number;
  /** Сумма post-tax удержаний (например, профсоюз). */
  postTaxDeductions: number;
  /** Разбивка по каждому extra-удержанию (для UI). */
  extraDeductionsApplied: SalaryDeductionLine[];
  incomeTax: number;
  net: number;
}

/**
 * Расчёт «зарплаты на руки» в Беларуси.
 *
 *   налогооблагаемая база = max(0, gross − стандартные вычеты − дет. вычеты
 *                                  − pre-tax extra-удержания)
 *   подоходный            = база × 13%
 *   ФСЗН (с работника)    = gross × 1%
 *   на руки               = gross − подоходный − ФСЗН
 *                           − pre-tax extra (уже учтены в базе → списываем)
 *                           − post-tax extra (профсоюз, ДМС, благотворительность)
 */
export function calcNetSalary(input: SalaryInput): SalaryResult {
  const {
    gross,
    children = 0,
    dependents = 0,
    applyStandardDeduction = true,
    extraDeductions = [],
  } = input;

  let deductions = 0;
  if (applyStandardDeduction && gross <= STD_DEDUCTION_INCOME_LIMIT) {
    deductions += STD_DEDUCTION;
  }
  if (children > 0) {
    if (children === 1) deductions += CHILD_DEDUCTION;
    else deductions += children * CHILD_DEDUCTION_TWO_PLUS;
  }
  if (dependents > 0) deductions += dependents * CHILD_DEDUCTION;

  const extraLines: SalaryDeductionLine[] = extraDeductions.map(d => {
    const amount = d.kind === 'percent'
      ? (gross * d.value) / 100
      : d.value;
    return {
      id: d.id,
      label: d.label,
      amount: Math.max(0, amount),
      taxable: Boolean(d.taxable),
    };
  });

  const pretaxDeductions = extraLines
    .filter(l => l.taxable)
    .reduce((s, l) => s + l.amount, 0);
  const postTaxDeductions = extraLines
    .filter(l => !l.taxable)
    .reduce((s, l) => s + l.amount, 0);

  const taxableBase = Math.max(0, gross - deductions - pretaxDeductions);
  const incomeTax = (taxableBase * INCOME_TAX_PCT) / 100;
  const fszn = (gross * FSZN_EMPLOYEE_PCT) / 100;
  const net = gross - incomeTax - fszn - pretaxDeductions - postTaxDeductions;

  return {
    gross,
    fszn,
    taxableBase,
    deductions,
    pretaxDeductions,
    postTaxDeductions,
    extraDeductionsApplied: extraLines,
    incomeTax,
    net,
  };
}

/** Inverse: какая «грязная» нужна для целевой «на руки»? */
export function grossFromNet(targetNet: number, opts: Omit<SalaryInput, 'gross'> = {}): number {
  // Binary search — нелинейно из-за порога льгот.
  let lo = 0,
    hi = targetNet * 3,
    mid = 0;
  for (let i = 0; i < 80; i++) {
    mid = (lo + hi) / 2;
    const { net } = calcNetSalary({ ...opts, gross: mid });
    if (Math.abs(net - targetNet) < 0.01) return mid;
    if (net < targetNet) lo = mid;
    else hi = mid;
  }
  return mid;
}

/* ------------------------------------------------------------------ */
/*                       Депозит — налог 13%                           */
/* ------------------------------------------------------------------ */

export interface DepositInput {
  amount: number;
  annualRatePct: number;
  termMonths: number;
  /**
   * Применять ли подоходный налог 13% на проценты по вкладам сроком
   * меньше года (для физлиц с 2024). Передавайте false для безотзывных
   * длинных депозитов и для валютных вкладов >= 2 лет.
   */
  taxApplies?: boolean;
  /** Капитализация процентов (true) или ежемесячная выплата (false). */
  capitalize?: boolean;
}

export interface DepositResult {
  totalInterestGross: number;
  totalTax: number;
  totalInterestNet: number;
  finalBalance: number;
}

export function calcDepositBY(input: DepositInput): DepositResult {
  const {
    amount,
    annualRatePct,
    termMonths,
    taxApplies = true,
    capitalize = true,
  } = input;
  const monthlyRate = annualRatePct / 100 / 12;

  let balance = amount;
  let interestGross = 0;
  for (let m = 1; m <= termMonths; m++) {
    const i = balance * monthlyRate;
    interestGross += i;
    if (capitalize) balance += i;
  }
  const tax = taxApplies ? (interestGross * INCOME_TAX_PCT) / 100 : 0;
  const interestNet = interestGross - tax;
  return {
    totalInterestGross: interestGross,
    totalTax: tax,
    totalInterestNet: interestNet,
    finalBalance: capitalize ? balance - tax : amount + interestNet,
  };
}

/* ------------------------------------------------------------------ */
/*                       ИП на УСН (5% / 3%)                           */
/* ------------------------------------------------------------------ */

export interface IpUsnInput {
  /** Годовая выручка ИП. */
  annualRevenue: number;
  /** 5% — без НДС, 3% — с НДС. */
  ratePct: 3 | 5;
  /** Фиксированный взнос ФСЗН в месяц (вводится пользователем). */
  fsznMonthly: number;
}

export interface IpUsnResult {
  usnTax: number;
  fsznTotal: number;
  totalLoad: number;
  net: number;
  effectivePct: number;
}

export function calcIpUsn(input: IpUsnInput): IpUsnResult {
  const { annualRevenue, ratePct, fsznMonthly } = input;
  const usnTax = (annualRevenue * ratePct) / 100;
  const fsznTotal = fsznMonthly * 12;
  const totalLoad = usnTax + fsznTotal;
  const net = annualRevenue - totalLoad;
  return {
    usnTax,
    fsznTotal,
    totalLoad,
    net,
    effectivePct: annualRevenue ? (totalLoad / annualRevenue) * 100 : 0,
  };
}

/* ------------------------------------------------------------------ */
/*                Налог на профессиональный доход (НПД)                */
/* ------------------------------------------------------------------ */

/**
 * Самозанятые физлица:
 *   10% — при работе с физлицами и иностранными организациями;
 *   20% — для дохода свыше 60 000 руб. в год (или сверхлимит).
 */
export const NPD_LOW_RATE_PCT = 10;
export const NPD_HIGH_RATE_PCT = 20;
export const NPD_HIGH_RATE_THRESHOLD = 60_000;

export interface NpdInput {
  annualRevenue: number;
}

export interface NpdResult {
  taxLow: number;
  taxHigh: number;
  total: number;
  net: number;
  effectivePct: number;
}

export function calcNpd(input: NpdInput): NpdResult {
  const r = input.annualRevenue;
  const low = Math.min(r, NPD_HIGH_RATE_THRESHOLD);
  const high = Math.max(0, r - NPD_HIGH_RATE_THRESHOLD);
  const taxLow = (low * NPD_LOW_RATE_PCT) / 100;
  const taxHigh = (high * NPD_HIGH_RATE_PCT) / 100;
  const total = taxLow + taxHigh;
  return {
    taxLow,
    taxHigh,
    total,
    net: r - total,
    effectivePct: r ? (total / r) * 100 : 0,
  };
}

/* ------------------------------------------------------------------ */
/*                       Отпускные / больничные                        */
/* ------------------------------------------------------------------ */

/**
 * Отпускные (упрощённо): среднедневной заработок × число календарных дней
 * отпуска. Среднедневной = доход за 12 мес / (12 × 29.7).
 */
export function calcVacationPay(
  totalEarningsLast12m: number,
  daysOff: number
): { avgDaily: number; payment: number } {
  const avgDaily = totalEarningsLast12m / (12 * 29.7);
  return { avgDaily, payment: avgDaily * daysOff };
}

File: flutter_app/lib/finance/by_tax.dart
// Belarus-specific tax calculators (port of src/lib/finance/byTax.ts).

/// Year-dependent НК РБ parameters. МЗП, base value and deduction thresholds
/// change almost every year, so they live in a table keyed by year — updating
/// is a one-line addition. Port of `BY_TAX_BY_YEAR` in `src/lib/finance/byTax.ts`.
///
/// ⚠️ Only 2024 values are confirmed. Add 2025/2026 from official sources
/// (Совмин postановления for МЗП/БВ and art. 209 НК РБ for deductions).
class ByTaxYearConstants {
  final double baseValueByn;
  final double minWageByn;
  final double stdDeductionIncomeLimit;
  final double stdDeduction;
  final double childDeduction;
  final double childDeductionTwoPlus;
  final double incomeTaxPct;
  final double fsznEmployeePct;

  const ByTaxYearConstants({
    required this.baseValueByn,
    required this.minWageByn,
    required this.stdDeductionIncomeLimit,
    required this.stdDeduction,
    required this.childDeduction,
    required this.childDeductionTwoPlus,
    required this.incomeTaxPct,
    required this.fsznEmployeePct,
  });
}

const Map<int, ByTaxYearConstants> byTaxByYear = {
  2024: ByTaxYearConstants(
    baseValueByn: 42,
    minWageByn: 626,
    stdDeductionIncomeLimit: 1054,
    stdDeduction: 174,
    childDeduction: 51,
    childDeductionTwoPlus: 97,
    incomeTaxPct: 13,
    fsznEmployeePct: 1,
  ),
};

/// Latest year with confirmed values — used by default.
const int currentTaxYear = 2024;

/// Returns constants for [year]: exact match → nearest earlier filled year →
/// latest available, so the calculator never breaks for an un-filled year.
ByTaxYearConstants getByTaxConstants([int year = currentTaxYear]) {
  final exact = byTaxByYear[year];
  if (exact != null) return exact;
  final years = byTaxByYear.keys.toList()..sort();
  final fallbackYear = years.reversed.firstWhere(
    (y) => y <= year,
    orElse: () => years.last,
  );
  return byTaxByYear[fallbackYear]!;
}

final ByTaxYearConstants _current = getByTaxConstants(currentTaxYear);

final double baseValueByn = _current.baseValueByn;
final double minWageByn = _current.minWageByn;
final double stdDeductionIncomeLimit = _current.stdDeductionIncomeLimit;
final double stdDeduction = _current.stdDeduction;
final double childDeduction = _current.childDeduction;
final double childDeductionTwoPlus = _current.childDeductionTwoPlus;
final double incomeTaxPct = _current.incomeTaxPct;
final double fsznEmployeePct = _current.fsznEmployeePct;

// NPD
const double npdLowRatePct = 10;
const double npdHighRatePct = 20;
const double npdHighRateThreshold = 60000;

class SalaryDeduction {
  final String id;
  final String label;
  final String kind; // 'percent' | 'fixed'
  final double value;
  final bool taxable;

  const SalaryDeduction({
    required this.id,
    required this.label,
    required this.kind,
    required this.value,
    this.taxable = false,
  });
}

class SalaryDeductionLine {
  final String id;
  final String label;
  final double amount;
  final bool taxable;

  const SalaryDeductionLine({
    required this.id,
    required this.label,
    required this.amount,
    required this.taxable,
  });
}

class SalaryResult {
  final double gross;
  final double fszn;
  final double taxableBase;
  final double deductions;
  final double pretaxDeductions;
  final double postTaxDeductions;
  final List<SalaryDeductionLine> extraDeductionsApplied;
  final double incomeTax;
  final double net;

  const SalaryResult({
    required this.gross,
    required this.fszn,
    required this.taxableBase,
    required this.deductions,
    required this.pretaxDeductions,
    required this.postTaxDeductions,
    required this.extraDeductionsApplied,
    required this.incomeTax,
    required this.net,
  });
}

SalaryResult calcNetSalary({
  required double gross,
  int children = 0,
  int dependents = 0,
  bool applyStandardDeduction = true,
  List<SalaryDeduction> extraDeductions = const [],
}) {
  double deductions = 0;
  if (applyStandardDeduction && gross <= stdDeductionIncomeLimit) {
    deductions += stdDeduction;
  }
  if (children > 0) {
    if (children == 1) {
      deductions += childDeduction;
    } else {
      deductions += children * childDeductionTwoPlus;
    }
  }
  if (dependents > 0) {
    deductions += dependents * childDeduction;
  }

  final extraLines = extraDeductions
      .map((d) => SalaryDeductionLine(
            id: d.id,
            label: d.label,
            amount:
                d.kind == 'percent' ? (gross * d.value) / 100 : d.value,
            taxable: d.taxable,
          ))
      .toList();

  final pretax =
      extraLines.where((l) => l.taxable).fold(0.0, (s, l) => s + l.amount);
  final postTax =
      extraLines.where((l) => !l.taxable).fold(0.0, (s, l) => s + l.amount);

  final taxBase = (gross - deductions - pretax).clamp(0.0, double.infinity);
  final tax = (taxBase * incomeTaxPct) / 100;
  final fszn = (gross * fsznEmployeePct) / 100;
  final net = gross - tax - fszn - pretax - postTax;

  return SalaryResult(
    gross: gross,
    fszn: fszn,
    taxableBase: taxBase,
    deductions: deductions,
    pretaxDeductions: pretax,
    postTaxDeductions: postTax,
    extraDeductionsApplied: extraLines,
    incomeTax: tax,
    net: net,
  );
}

double grossFromNet(double targetNet,
    {int children = 0,
    int dependents = 0,
    bool applyStandardDeduction = true,
    List<SalaryDeduction> extraDeductions = const []}) {
  if (targetNet <= 0) return 0;

  double netAt(double gross) => calcNetSalary(
        gross: gross,
        children: children,
        dependents: dependents,
        applyStandardDeduction: applyStandardDeduction,
        extraDeductions: extraDeductions,
      ).net;

  // Expand the upper bound until it actually brackets the target (percentage
  // deductions can push the effective net-to-gross ratio well below 1), rather
  // than assuming 3× is always enough.
  double lo = 0;
  double hi = targetNet <= 0 ? 1 : targetNet * 1.5;
  var guard = 0;
  while (netAt(hi) < targetNet && guard < 60) {
    lo = hi;
    hi *= 2;
    guard++;
  }
  // No gross yields the requested net (e.g. deductions exceed 100%).
  if (netAt(hi) < targetNet) return double.nan;

  double mid = hi;
  for (int i = 0; i < 100; i++) {
    mid = (lo + hi) / 2;
    final net = netAt(mid);
    if ((net - targetNet).abs() < 0.005) return mid;
    if (net < targetNet) {
      lo = mid;
    } else {
      hi = mid;
    }
  }
  return mid;
}

// --- Deposit BY (13% tax on short-term) ---

class DepositResult {
  final double totalInterestGross;
  final double totalTax;
  final double totalInterestNet;
  final double finalBalance;

  const DepositResult({
    required this.totalInterestGross,
    required this.totalTax,
    required this.totalInterestNet,
    required this.finalBalance,
  });
}

DepositResult calcDepositBY({
  required double amount,
  required double annualRatePct,
  required int termMonths,
  bool taxApplies = true,
  bool capitalize = true,
}) {
  if (termMonths <= 0 || amount <= 0) {
    return const DepositResult(
      totalInterestGross: 0,
      totalTax: 0,
      totalInterestNet: 0,
      finalBalance: 0,
    );
  }
  final monthlyRate = annualRatePct / 100 / 12;
  double balance = amount;
  double interestGross = 0;
  for (int m = 1; m <= termMonths; m++) {
    final i = balance * monthlyRate;
    interestGross += i;
    if (capitalize) balance += i;
  }
  final tax = taxApplies ? (interestGross * incomeTaxPct) / 100 : 0.0;
  final interestNet = interestGross - tax;
  return DepositResult(
    totalInterestGross: interestGross,
    totalTax: tax,
    totalInterestNet: interestNet,
    finalBalance: capitalize ? balance - tax : amount + interestNet,
  );
}

// --- IP USN (5% / 3%) ---

class IpUsnResult {
  final double usnTax;
  final double fsznTotal;
  final double totalLoad;
  final double net;
  final double effectivePct;

  const IpUsnResult({
    required this.usnTax,
    required this.fsznTotal,
    required this.totalLoad,
    required this.net,
    required this.effectivePct,
  });
}

/// ⚠️ Historical/reference only. The УСН regime for ИП was abolished in
/// Belarus from 2023; new activity uses ОСН/подоходный or НПД instead. Kept for
/// modelling legacy periods — do not present as a current option.
IpUsnResult calcIpUsn({
  required double annualRevenue,
  required int ratePct,
  required double fsznMonthly,
}) {
  final usnTax = (annualRevenue * ratePct) / 100;
  final fsznTotal = fsznMonthly * 12;
  final totalLoad = usnTax + fsznTotal;
  final net = annualRevenue - totalLoad;
  return IpUsnResult(
    usnTax: usnTax,
    fsznTotal: fsznTotal,
    totalLoad: totalLoad,
    net: net,
    effectivePct: annualRevenue > 0 ? (totalLoad / annualRevenue) * 100 : 0,
  );
}

// --- NPD (self-employed) ---

class NpdResult {
  final double taxLow;
  final double taxHigh;
  final double total;
  final double net;
  final double effectivePct;

  const NpdResult({
    required this.taxLow,
    required this.taxHigh,
    required this.total,
    required this.net,
    required this.effectivePct,
  });
}

NpdResult calcNpd({required double annualRevenue}) {
  final r = annualRevenue;
  final low = r < npdHighRateThreshold ? r : npdHighRateThreshold;
  final high = (r - npdHighRateThreshold).clamp(0.0, double.infinity).toDouble();
  final taxLow = (low * npdLowRatePct) / 100;
  final taxHigh = (high * npdHighRatePct) / 100;
  final total = taxLow + taxHigh;
  return NpdResult(
    taxLow: taxLow,
    taxHigh: taxHigh,
    total: total,
    net: r - total,
    effectivePct: r > 0 ? (total / r) * 100 : 0,
  );
}

// --- Vacation pay ---

/// Vacation pay via the average-daily-earnings method.
///
/// Uses the 29.6 average-monthly-days coefficient (Постановление Совмина РБ
/// № 1290 as amended); the exact figure is periodically revised, so treat the
/// result as an approximation for planning rather than a payroll-exact number.
({double avgDaily, double payment}) calcVacationPay(
    double totalEarningsLast12m, int daysOff) {
  final avgDaily = totalEarningsLast12m / (12 * 29.6);
  return (avgDaily: avgDaily, payment: avgDaily * daysOff);
}
